ocaml/typing/subst.ml

```diff
@@ -765,6 +765,8 @@ and subst_lazy_modtype scoping s = function
                   subst_lazy_modtype scoping (add_module id (Pident id') s) res)
   | Mty_alias p ->
       Mty_alias (module_path s p)
+  | Mty_strengthen (mty, p, a) ->
+      Mty_strengthen (subst_lazy_modtype scoping s mty, module_path s p, a)
```

Comment:
No, we quite explicitly don't want to update `a` since we want it to be stable under substitution. It indicates whether or not the strengthening in question is aliasable. It can only be aliasable if the module in question is not a functor argument - but it's fine for it to be non-aliasable in either case. So the only problematic case would be if we substituted a functor parameter for a real module - then, strengthening could be aliasable before but not after. However, it is a meta property of the type checker that this never happens. It's not a new requirement, either, since we already can't substitute functor arguments into `Mty_alias`.

Reply:
> It can only be aliasable if the module in question is not a functor argument - but it's fine for it to be non-aliasable in either case. So the only problematic case would be if we substituted a functor parameter for a real module

I'm probably just being dense, but I think the above example demonstrates the problematic case.  The strengthening is aliasable when it is declared in `S2`, because `M1` is not a functor argument.  But then we do substitute a functor argument in for `M1`, and the strengthening remains aliasable.

That is, I think `S2 with module M1 := X` will contain `Mty_strengthen(S1, X, Aliasable)`, and I think you are saying that can never happen because `X` is a functor argument.

Reply:
This is analgous to:
```ocaml
module type S1 = sig
    type t
  end

module type S2 = sig
  module M1 : S1
  module M2 = M1
end

module F(X : S1) : S2 with module M1 := X = struct
  module M1 = X
  module M2 = M1
end
```
which does indeed create an invalid module type, but then gives an appropriate error when attempting the inclusion:
```ocaml
Error: Signature mismatch:
       Modules do not match:
         sig module M1 : sig type t = X.t end module M2 = M1 end
       is not included in
         sig module M2 = X end
       In module M2:
       Module X cannot be aliased
```

I actually don't think it's possible to fully prevent aliases to functor parameters, despite it supposedly being an invariant in the compiler. The whole approach is broken on a fundamental level. However, I don't know of any cases where this causes actual unsoundness. The solution is to get rid of this invariant and support aliases of functor parameters (and by extension functor applications).

This PR isn't making any of this any worse though. We should include your example in the test suite to check we get a sensible error like the one above.

We should also add this case:
```ocaml
module type S1 = sig
    module M : sig type t end
end

module type S2 = sig
  module M1 : S1
  module M2 : S1 with M1
end

module F(X : S2) = struct
  include X
end
```
which I think is probably producing an alias to a functor parameter in the current PR, when the functor should instead produce a module with type:
```ocaml
sig
      module M1 : sig module M : sig type t = X.M1.M.t end end
      module M2 : sig module M : sig type t = M1.M.t end end
end
```
This is because the PR currently does not update `Mtype.remove_aliases_mty`.

Reply:
Firstly, apologies for not reading your example carefully. There are a couple of things going on here.

1. You are indeed correct that the substitution in the implementation of destructive `with` breaks this invariant. It also can end up with `Mty_alias X` where `X` is a functor argument - such programs then get rejected during signature matching in `Includemod`. Reusing substitution for implementing destructive `with` is probably not entirely optimal as it's not "real" substitution - hence why I never think about it when talking about substitution in general.
2. When typing destructive `with` we expand all strengthening nodes which could possibly be affected because of the weird substitution. This means that there wouldn't be a `Mty_strengthen (S1, M1, Aliasable)` node to substitute `X` for `M1` into because we would have expanded it. 
3. Unfortunately, there is a bug in `Mtype.expand_to` which means it doesn't expand enough. So it is indeed currently possible to construct aliases to functor arguments this way but that's because of the bug in `expand_to`, not because of anything in `Subst`. That's easy to fix, though.
4. Equally unfortunately, we currently `assert false` if `S1` is abstract because we can't expand the node. This is also a bug - I didn't realise this can actually happen but it obviously can. This one also has a couple of easy fixes but alas, I'm quite unsure what the semantics should be in this case. I'll ponder more and then push the two fixes together.

You example was extremely helpful in identifying the two bugs, thanks a lot!

Reply:
> which I think is probably producing an alias to a functor parameter in the current PR

Doesn't seem to be the case. In any event, your point about `remove_aliases_mty` is well taken.

Reply:
> This is because the PR currently does not update `Mtype.remove_aliases_mty`.

Actually, after paging that back in (I really should add a comment): `remove_aliases_mty` has two call sites.

1. From `scrape_for_type_of` for things with the `remove_aliases` attribute. IIRC, we agreed to not bother supporting it. Should I remove it?
2. From `scrape_for_functor_arg`. Here, we always scrape the type during the traversal so strengthening nodes will be resolved except for abstract types. I'm not quite sure what it means to remove aliases from a strengthened abstract type - I suspect it's a noop.

So I don't think it actually needs updating.

Reply:
I think that `remove_aliases` should change the aliasable bool from `true` to `false` when the path of the module is not in the excluded by the `exclude` function. This will only matter in the abstract case due to the scraping, but without it it should be possible to create an alias to part of a functor parameter.

Reply:
> Doesn't seem to be the case.

Ah, I'd missed the scrapping. I guess it would need to be something like:
```ocaml
module type Type = sig module type S end

module Foo (X : Type) = struct

  module type S = sig
    module M1 : X.S
    module M2 : X.S with M1
  end

  module F(X : S2) = struct
    include X.M2
  end
end

module Sub = struct
  module type S = sig
    module M : sig val x : int end
  end
end

module Bar = Foo(Sub)
```
You might have to play around a bit to get it, but I think roughly that should work.

Reply:
This is all done now, I think, include `remove_aliases`. In the case of destructive substitution, we rely on the inclusion check just like we do for module aliases so no further expanding is necessary.